flips/20220817-extensibility.md

```diff
@@ -0,0 +1,423 @@
+is run automatically when the extension is removed from a value (before the removal occurs). These functions exist to perform any necessary setup
+and teardown for the extended type that requires using values from the base type, and thus cannot be performed in `init` or `destroy`. It is also recommended that 
+users check any pre-conditions or post-conditions they would like to hold before or after the extended type is created in these functions. 
+
+If a resource type with attached extensions is `destroy`ed, the extensions will be implicitly `remove`d and then `destroy`ed. For reasons that will be further
+detailed in the Extended Types section below, there are no guarantees made about the order in which the `remove` and `destroy` methods will be run in the case
+of destroying a resource type with multiple extensions attached, other than that a) `destroy` on the base type will always be the last method run, and b) the
+`remove` method for an extension will always be run before the `destroy` method of that same extension. If developers wish to enforce that `remove`s and `destroy`s
+run in a specific order, it is recommended that they explicitly `remove` extensions from a resource and `destroy` them, rather than `destroy`ing the entire extended
+resource type all at once. 
+
+Some may wonder why `init` and `attach` are separate functions (and have a similar question about `destroy` and `remove`). The primary reason for this
+is to allow extensions to be re-used across multiple instances of the same base type. By separating `init` and `attach` (and `destroy` and `remove`), 
+we can have separate code that runs once when the extension is created (destroyed), and then multiple times each time the extension is attached to (removed from) 
+a base type. A motivating example for this would be a hat for a CryptoKitty: by separating the code to initialize the hat and the code to "equip" a Kitty with the 
+hat, the hat can be added and removed from various Kitties, as well as sold and traded independently of the Kitties themselves.  
+
+### Extended Types
+
+The `with` type operator can be used to describe the type of an extended composite. Hence, given some resource `R` and an extension `E` defined as such:
+
+```cadence
+resource R {}
+extension E for R {}
+```
+
+The type `@R with @E` describes the type of `R` extended with `E`. An extension can only appear on the right side of the `with` type operator 
+when it is a valid extension for the type on the left. 
+
+The extension type itself also can be referenced with just `E`, but no operations can be performed on a value of type `E` other than to 
+move it around or to attach it to a value using `extend` syntax (see below). That is to say, a method or field defined in the declaration of `E` 
+exists on values of type `@R with @E`, not on values of type `E`. 
```

Comment:
I assume that function calls on just `E` will be statically rejected? In the case where dynamic dispatch is involved, e.g. an instance of `E` is assigned to a variable of type `{I}`, we will also need to dynamically reject the direct function calls. 
Maybe add that to the proposal, if it is the planned behaviour. If not, how else could we handle this?

This limitation will also be a bit restrictive. For example, given that  an extension could be an NFT, it would be impossible to get metadata of it in the detached state.

Reply:
> `an instance of E is assigned to a variable of type {I}`

At the moment this isn't a concern because extensions cannot implement interfaces in their detached state, so there would never be any dynamic dispatch to worry about. 

However, this is also unlikely to be a relevant concern for much longer; based on the discussion in the last meeting about extensibility, I think the sentiment is that extensions should not themselves be resources with value, and should not be usable independently of the thing they are attached to. I.e. removing an extension should also destroy it. Instead the extension would be a "manager" for the valuable resource, perhaps through the indirection of a reference or capability.